src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,241 @@
+#pragma once
+#include <chrono>
+#include <thread>
+#include <condition_variable>
+#include "rgw_common.h"
+
+#if !defined(dout_subsys)
+#define dout_subsys ceph_subsys_rgw
+#define def_dout_subsys
+#endif
+
+#define dout_context g_ceph_context
+
+class QosDatastruct {
+  enum {
+    RGW_QOS_OPS,
+    RGW_QOS_BW,
+    RGW_QOS_ROUND_OPS,
+    RGW_QOS_ROUND_BYTES
+  };
+  struct qos_entry {
+    // 3 is the enum size
+    std::atomic_int64_t read[5] = {0};
+    std::atomic_int64_t write[5] = {0};
+    int64_t ts = ceph::coarse_real_clock::now().time_since_epoch().count();
+    bool first_run = true;
+    std::mutex ts_lock;
+  };
+  const size_t map_size = 2000000; // will create it with the closest upper prime number
+  std::mutex insert_lock;
+  std::condition_variable *cv;
+  std::atomic_bool *replacing;
+  typedef std::unordered_map<std::string, qos_entry> hash_map;
+  hash_map qos_entries{map_size};
+  bool is_read_op(const char *method) {
+    if((strcmp(method,"GET") == 0) || (strcmp(method,"HEAD") == 0))
+      return true;
+    return false;
+  }
+    // find or create an entry, and return its iterator
+  auto find_or_create(const std::string& key) {
+    if(qos_entries.size() > 0.9 * map_size && *replacing == false)
+    {
+        *replacing = true;
+        cv->notify_all();
+    }
+    auto ret = qos_entries.find(key);
+    if(ret == qos_entries.end())
+    {
+      insert_lock.lock();
+      ret = qos_entries.emplace(std::piecewise_construct,
+                                 std::forward_as_tuple(key),
+                                 std::forward_as_tuple()).first;
+      insert_lock.unlock();
```

Comment:
i don't see how it could be safe to only hold the lock over `emplace()`, but not over `find()` and `size()`